[Jordan Soh Jing Ren] iP

jorrdansoh opened the pull request:
# Duke
It
* Can talk
* Responds to your queries

All you need to do is
1. Download it
2. Run it
3. Add your tasks

Run from `main`:
```java
public static void main(String[] args) {
        Duke duke = new Duke();
        duke.run();
    }
```

Features
- [x] Find tasks
- [ ] GUI
- [ ] Personality

:shipit:

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

*Best of all*, it is **FREE**

Commit:
Add Gradle support

jorrdansoh committed:
Added UI and implemented level-1 for Thomas

jorrdansoh committed:
Added ability to store commands into list and display them

jorrdansoh committed:
Seperated commands into multiple classes, added parser to parse user input, added ability to mark tasks.

jorrdansoh committed:
Added three new types of tasks

jorrdansoh committed:
Added automated UI testing

jorrdansoh committed:
Added error handling and ability to delete tasks

jorrdansoh committed:
Fixed some minor Documentation issues

jorrdansoh committed:
Added more error handling

jorrdansoh committed:
Add ability to save to hard disk

jorrdansoh committed:
Add ability to parse dates

jorrdansoh committed:
Resolve merge conflict

jorrdansoh committed:
Refactor some code

jorrdansoh committed:
Add more OOP

jorrdansoh committed:
Separate classes into respective packages

jorrdansoh committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip into A-Gradle

* 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip:
  Add Gradle support

jorrdansoh committed:
Add Gradle support and JUnit tests

jorrdansoh committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

* commit '556af3f47a96b32898ab4cdbd65b16486a4871e8':
  Add Gradle support

jorrdansoh committed:
Merge branch 'master' into A-Gradle

* master:

jorrdansoh committed:
Reorder imports

jorrdansoh committed:
Merge branch 'master' into A-Gradle

* master:
  Reorder imports

jorrdansoh committed:
Reorder imports

jorrdansoh committed:
Add JavaDoc comments

jorrdansoh committed:
Add CheckStyle to follow coding standard

jorrdansoh committed:
Add a way for users to find tasks by keyword

jorrdansoh committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

jorrdansoh committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Add CheckStyle to follow coding standard

# Conflicts:
#	src/main/java/dwuke/Parser.java
#	src/main/java/dwuke/task/TaskList.java

jorrdansoh committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add a way for users to find tasks by keyword

# Conflicts:
#	src/main/java/dwuke/Parser.java
#	src/main/java/dwuke/task/Task.java
#	src/main/java/dwuke/task/TaskList.java

jorrdansoh committed:
Update UI test and refactor code

jorrdansoh committed:
Change Dwuke to Duke

jorrdansoh committed:
Fix typo

jorrdansoh committed:
Git ignore duke.txt

jorrdansoh committed:
Fix typo

jorrdansoh committed:
Make some class variables private

jorrdansoh committed:
Add Checkstyle

jorrdansoh committed:
Add GUI for Duke

jorrdansoh committed:
Fix style

jorrdansoh committed:
Merge branch 'Level-10'

* Level-10:
  Fix style

jorrdansoh committed:
Add credit for reused code

jorrdansoh committed:
Add credit for reused code

Piyotato commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    @Override
+    public String toString() {
+        String result = "";
```
Consider using StringBuilder instead of adding strings?

jorrdansoh replied:
That's a good idea! I'll consider using that, thanks.

Piyotato commented on src/main/java/duke/Date.java:
```diff
@@ -0,0 +1,44 @@
+     * @param localDate The LocalDate object to convert.
+     * @return A String representing the LocalDate.
+     */
+    public static String displayFormat(LocalDate localDate) {
```
Consider naming this method in a way that insinuates that it converts the local date _to_ the display format? Maybe `getDisplayFormat`?

jorrdansoh replied:
That's a good suggestion, thanks :)

Piyotato commented on src/main/java/duke/Date.java:
```diff
@@ -0,0 +1,44 @@
+     * @param localDate The LocalDate object to convert.
+     * @return A String representing the LocalDate.
+     */
+    public static String storageFormat(LocalDate localDate) {
```
Consider naming this method in a way that insinuates that it converts the local date _to_ the display format? Maybe `getStorageFormat`?

jorrdansoh replied:
That's a good suggestion, thanks :)

ruihan00 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * This class encapsulates a mark command from the user.
+ */
+public class MarkCommand extends Command {
+    public static final String COMMAND_WORD = "mark";
+
+    private TaskList taskList;
+    private int pos;
+
+    /**
+     * Creates a MarkCommand with the given TaskList and position.
+     *
+     * @param taskList The TaskList.
+     * @param pos      The position.
+     * @throws DukeException If the position is not in the TaskList.
+     */
+    public MarkCommand(TaskList taskList, int pos) throws DukeException {
+        if (pos < 0 || pos > taskList.size() - 1) {
+            throw new DukeException("The list doesn't have a task with that number.");
+        }
+        this.taskList = taskList;
+        this.pos = pos;
```
interesting

ruihan00 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * This class encapsulates Duke and its functionalities.
+ */
+public class Duke {
+    private final Storage storage;
+    private TaskList taskList;
+
+    /**
+     * Creates a new Duke to chat with.
+     *
+     * @throws DukeException If there is a problem retrieving the saved tasks.
+     */
+    public Duke() throws DukeException {
+        this.storage = new Storage();
+        this.taskList = this.storage.load();
+    }
+
+    /**
+     * Parses the given text using the Parser.
+     *
+     * @param text The text to parse.
+     * @return The corresponding command after parsing the text.
+     */
+    public Command parseText(String text) {
+        return Parser.parseText(text, this.taskList);
+    }
+
+    /**
+     * Saves the TaskList to the hard disk.
```
very neat

ruihan00 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+     * @param pos      The position.
+     * @throws DukeException If the position is not in the TaskList.
+     */
+    public DeleteCommand(TaskList taskList, int pos) throws DukeException {
```
perhaps can change public to protected for some of the methods

jorrdansoh replied:
That's a good suggestion, thanks!

ruihan00 reviewed (commented):
very neat code!

jorrdansoh committed:
Add Java assertions

Classes (e.g. MainWindow) have no assertions to document
important assumptions.

Adding Java assertions in various places to check that various objects
are not null.

Asserting that they are not null is important as they should hold true
at various points in the code.

jorrdansoh committed:
Improve code quality

parseText method in Parser class exceeds 30 LoC.

Extracting the switch statement in parseText to another new method
called getCommand.

Long switch statement in parseText reduces readability, thus it was
extracted out to another method.

jorrdansoh committed:
Merge pull request #1 from jorrdansoh/branch-A-Assertions

Add Java assertions

jorrdansoh committed:
Merge branch 'master' of github.com:jorrdansoh/ip

* 'master' of github.com:jorrdansoh/ip:
  Add Java assertions

jorrdansoh committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add Java assertions

# Conflicts:
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java

jorrdansoh committed:
Merge pull request #2 from jorrdansoh/branch-A-CodeQuality

Improve code quality

jorrdansoh committed:
Merge branch 'master' of github.com:jorrdansoh/ip

* 'master' of github.com:jorrdansoh/ip:
  Improve code quality

jorrdansoh committed:
Fix styling

jorrdansoh committed:
Add sort command

As a user adds tasks, their list becomes cluttered and disorganized.

Adding the sort command gives them the ability to sort their task list,
in the following order: deadlines and events in chronological order,
then tasks in alphabetical order.

Sorting deadlines and events in chronological order is beneficial to the
user as it lets them see the more urgent tasks first. Undated todos are
less urgent than dated deadlines and events, hence they are
sorted last.

jorrdansoh committed:
Fix JavaDoc wordings

jorrdansoh committed:
Merge branch 'branch-C-Sort'

* branch-C-Sort:
  Fix JavaDoc wordings

jorrdansoh committed:
Add reuse credit and rename methods

jorrdansoh committed:
Add product screenshot and UG template

jorrdansoh committed:
Complete User Guide

jorrdansoh committed:
Add title to GUI

jorrdansoh committed:
Update README.md
